### PR TITLE
Added 370,371,372,373,374,375,376,379

### LINKED ISSUE #157: Error code 370 - Child in independent living should be at least 15.
#### Error code
370
#### Description to display
Child in independent living should be at least 15.
#### DfE Long-form description of error
Where the placement code is P2 the date at the start of the placement should be the child's 15th birthday or over.  Check that the date of birth and the placement code are correct. If so contact DfE.
#### DfE suggested coding
Where `<PL>` = 'P2' then `<DOB>` should be >`= 15 years prior to `<DECOM>`

For further info, refer to the DfE validation document [here](https://assets.publishing.service.gov.uk/government/uploads/system/uploads/attachment_data/file/974999/Children_looked_after_data_collection_2020_to_2021_validation_checks_-_version_1.3.pdf)

### LINKED ISSUE #158: Error code 371 - Child in semi- independent living accommodation not subject to children's homes regulations should be at least 14.
#### Error code
371
#### Description to display
Child in semi- independent living accommodation not subject to children's homes regulations should be at least 14.
#### DfE Long-form description of error
Where the placement code is H5 the date at the start of the placement should be the child's 14th birthday or over.  Check that the date of birth and the placement code are correct. If so contact DfE.
#### DfE suggested coding
Where `<PL>` = 'H5' then `<DOB>` should be >`= 14 years prior to `<DECOM>`

For further info, refer to the DfE validation document [here](https://assets.publishing.service.gov.uk/government/uploads/system/uploads/attachment_data/file/974999/Children_looked_after_data_collection_2020_to_2021_validation_checks_-_version_1.3.pdf)

### LINKED ISSUE #159: Error code 372 - Child in youth custody or prison should be at least 10.
#### Error code
372
#### Description to display
Child in youth custody or prison should be at least 10.
#### DfE Long-form description of error
Where the placement code is R5, the date at the start of the placement should be the child's 10th birthday or over.  Check that the date of birth and the placement code are correct.
#### DfE suggested coding
Where `<PL>` = 'R5' then `<DOB>` should be >`= 10 years prior to `<DECOM>`

For further info, refer to the DfE validation document [here](https://assets.publishing.service.gov.uk/government/uploads/system/uploads/attachment_data/file/974999/Children_looked_after_data_collection_2020_to_2021_validation_checks_-_version_1.3.pdf)

### LINKED ISSUE #160: Error code 373 - Child placed in a school should be at least 4 years old.
#### Error code
373
#### Description to display
Child placed in a school should be at least 4 years old.
#### DfE Long-form description of error
Where the placement code is S1 the date at the start of the placement should be the child's 4th birthday or over.  Check that the date of birth and the placement code are correct.
#### DfE suggested coding
Where `<PL>` = 'S1' then `<DOB>` should be >`= 4 years prior to `<DECOM>`

For further info, refer to the DfE validation document [here](https://assets.publishing.service.gov.uk/government/uploads/system/uploads/attachment_data/file/974999/Children_looked_after_data_collection_2020_to_2021_validation_checks_-_version_1.3.pdf)

### LINKED ISSUE #161: Error code 374 - Child in residential employment should be at least 14 years old.
#### Error code
374
#### Description to display
Child in residential employment should be at least 14 years old.
#### DfE Long-form description of error
Where the placement code is P3 the date at the start of the placement should be the child's 14th birthday or over.  Check that the date of birth and the placement code are correct.
#### DfE suggested coding
Where `<PL>` = 'P3' then `<DOB>` should be >`= 14 years prior to `<DECOM>`

For further info, refer to the DfE validation document [here](https://assets.publishing.service.gov.uk/government/uploads/system/uploads/attachment_data/file/974999/Children_looked_after_data_collection_2020_to_2021_validation_checks_-_version_1.3.pdf)

### LINKED ISSUE #162: Error code 375 - Hospitalisation coded as a temporary placement exceeds six weeks.
#### Error code
375
#### Description to display
Hospitalisation coded as a temporary placement exceeds six weeks.
#### DfE Long-form description of error
Placement T1 is only allowed for periods in hospital for the treatment of injury or acute illness and is therefore highly unlikely to exceed 42 days' duration.  Duration in days should be calculated by taking the start date away from the end date.  Check the reason the child has gone into hospital. If it is for the treatment of a chronic condition, the appropriate placement code is R2.  If after checking the record is correct as submitted contact DfE.
#### DfE suggested coding
Where `<PL>` = 'T1' then `<DEC>` should be `<= 42 days of `<DECOM>`

For further info, refer to the DfE validation document [here](https://assets.publishing.service.gov.uk/government/uploads/system/uploads/attachment_data/file/974999/Children_looked_after_data_collection_2020_to_2021_validation_checks_-_version_1.3.pdf)

### LINKED ISSUE #163: Error code 376 - Temporary placements coded as being due to holiday of usual foster carer(s) cannot exceed three weeks.
#### Error code
376
#### Description to display
Temporary placements coded as being due to holiday of usual foster carer(s) cannot exceed three weeks.
#### DfE Long-form description of error
Where the placement code is T3 the validation check will fail if the period of placement lasts longer than 21 days.  Duration in days should be calculated by taking the start date away from the end date. If a temporary placement starts on 1 June and ends on 22 June check will pass; if a temporary placement starts on 1 June and ends on 23 June check will fail.  To qualify as a temporary placement, the holiday of the usual foster carer(s) cannot exceed three weeks.
#### DfE suggested coding
Where `<PL>` = 'T3' then `<DEC>` should be `<= 21 days of `<DECOM>`

For further info, refer to the DfE validation document [here](https://assets.publishing.service.gov.uk/government/uploads/system/uploads/attachment_data/file/974999/Children_looked_after_data_collection_2020_to_2021_validation_checks_-_version_1.3.pdf)

### LINKED ISSUE #166: Error code 379 - Temporary placements for unspecified reason (placement code T4) cannot exceed seven days.
#### Error code
379
#### Description to display
Temporary placements for unspecified reason (placement code T4) cannot exceed seven days.
#### DfE Long-form description of error
Where the placement code is T4 the check will fail if the period of placement lasts longer than 7 days.  Duration in days should be calculated by taking the start date away from the end date. Temporary placement starts on 1 June and ends on 8 June check will pass; if temporary placement starts on 1 June and ends on 9 June check will fail.
#### DfE suggested coding
Where `<PL>` = 'T4' then `<DEC>` should be `<= 7 days of `<DECOM>`

For further info, refer to the DfE validation document [here](https://assets.publishing.service.gov.uk/government/uploads/system/uploads/attachment_data/file/974999/Children_looked_after_data_collection_2020_to_2021_validation_checks_-_version_1.3.pdf)

### PR DESCRIPTION
**Pull request by PhilDeakin**
Added 370, 371, 372, 373, 374, 375, 376, 379

### CODE REVIEW

File: validator903/validators.py [hunk 1]
```diff
@@ -2419,3 +2419,76 @@ def validate_503E():
...
+        else:
+            epi = dfs['Episodes']
+            hea = dfs['Header']
+            hea['DOB'] = pd.to_datetime(hea['DOB'], format='%d/%m/%Y', errors='coerce')
+            epi['DECOM'] = pd.to_datetime(epi['DECOM'], format='%d/%m/%Y', errors='coerce')
+            epi['DEC'] = pd.to_datetime(epi['DEC'], format='%d/%m/%Y', errors='coerce')
+            epi.reset_index(inplace=True)
+            epi_p2 = epi[epi['PLACE'] == Gen_370_dict[subval]['P_Code']]
+            merged_e = epi_p2.merge(hea, how='inner', on='CHILD')
+            if subval == '370' or subval == '371' or subval == '372' or subval == '373' or subval == '374':
```
Comment (PhilDeakin) [inline validator903/validators.py]: I presume there's an "isin" function or something equivalent to do this, but I got bored looking for it.

Comment (PhilDeakin) [reply validator903/validators.py]: Amended

File: validator903/validators.py [hunk 2]
```diff
@@ -2419,3 +2419,76 @@ def validate_503E():
...
+            hea = dfs['Header']
+            hea['DOB'] = pd.to_datetime(hea['DOB'], format='%d/%m/%Y', errors='coerce')
+            epi['DECOM'] = pd.to_datetime(epi['DECOM'], format='%d/%m/%Y', errors='coerce')
+            epi['DEC'] = pd.to_datetime(epi['DEC'], format='%d/%m/%Y', errors='coerce')
+            epi.reset_index(inplace=True)
+            epi_p2 = epi[epi['PLACE'] == Gen_370_dict[subval]['P_Code']]
+            merged_e = epi_p2.merge(hea, how='inner', on='CHILD')
+            if subval in  ['370', '371', '372', '373', '374']:
+                error_mask = merged_e['DECOM'] < (merged_e['DOB'] +
+                pd.offsets.DateOffset(years=Gen_370_dict[subval]['Y_gap']))
```
Comment (dezog) [inline validator903/validators.py]: Optional change: for style, better to indent this line so it's obvious it's a continuation of the previous one

File: validator903/validators.py [hunk 3]
```diff
@@ -2419,3 +2419,76 @@ def validate_503E():
...
+        else:
+            epi = dfs['Episodes']
+            hea = dfs['Header']
+            hea['DOB'] = pd.to_datetime(hea['DOB'], format='%d/%m/%Y', errors='coerce')
+            epi['DECOM'] = pd.to_datetime(epi['DECOM'], format='%d/%m/%Y', errors='coerce')
+            epi['DEC'] = pd.to_datetime(epi['DEC'], format='%d/%m/%Y', errors='coerce')
+            epi.reset_index(inplace=True)
+            epi_p2 = epi[epi['PLACE'] == Gen_370_dict[subval]['P_Code']]
+            merged_e = epi_p2.merge(hea, how='inner', on='CHILD')
+            if subval in  ['370', '371', '372', '373', '374']:
```
Comment (dezog) [inline validator903/validators.py:2782]: Yep, well done!

File: validator903/validators.py [hunk 4]
```diff
@@ -2419,3 +2419,76 @@ def validate_503E():
...
+            return {}
+        else:
+            epi = dfs['Episodes']
+            hea = dfs['Header']
+            hea['DOB'] = pd.to_datetime(hea['DOB'], format='%d/%m/%Y', errors='coerce')
+            epi['DECOM'] = pd.to_datetime(epi['DECOM'], format='%d/%m/%Y', errors='coerce')
+            epi['DEC'] = pd.to_datetime(epi['DEC'], format='%d/%m/%Y', errors='coerce')
+            epi.reset_index(inplace=True)
+            epi_p2 = epi[epi['PLACE'] == Gen_370_dict[subval]['P_Code']]
+            merged_e = epi_p2.merge(hea, how='inner', on='CHILD')
```
Comment (dezog) [inline validator903/validators.py:2780]: At this point, I think we should do `merged_e = merged_e.dropna(subset=['DECOM', 'DEC', 'DOB'])`, so this error doesn't trigger where there are missing/invalid values. Does that make sense?

File: tests/test_validators.py [hunk 5]
```diff
@@ -1714,3 +1714,140 @@ def test_validate_503F():
     result = error_func(fake_dfs)
 
     assert result == {'Episodes': [0, 5]}
+
+def test_validate_370():
+    fake_epi = pd.DataFrame({
+        'CHILD': ['111', '222', '333', '444', '555', '666'],
```
Comment (dezog) [inline tests/test_validators.py:1925]: Optional change: add pd.NA values to the tests to make sure that's working as expected.

**Review by dezog — COMMENTED** [review]
Nice work again! Almost two whole 'packs' from the spreadsheet in one commit 🤯 

image

Only one necessary change I can see - adding a dropna before comparing the values, check out the comment below

File: validator903/validators.py (hunk 2 above)
Comment (PhilDeakin) [reply validator903/validators.py]: Changed

File: validator903/validators.py (hunk 4 above)
Comment (PhilDeakin) [reply validator903/validators.py:2780]: Added

File: tests/test_validators.py (hunk 5 above)
Comment (PhilDeakin) [reply tests/test_validators.py:1925]: Added a few.

**Review by dezog — APPROVED** [review]
Fantastic, all present & correct - merging this now 🙂

File: validator903/validators.py (hunk 4 above)
Comment (dezog) [reply validator903/validators.py:2780]: I don't think this change made it in

**Review by dezog — CHANGES_REQUESTED** [review]
Sorry, I missed something - looks like the dropna line didnt make the commit, please could you double check that one?

**Comment by PhilDeakin** [comment]
Corrected now... (I hope) @dezog

**Review by dezog — APPROVED** [review]
Awesome, cheers Phil! As per earlier email, in future I'll tend to just make quick little tweaks like this myself, and just leave a comment if it's something that'd be useful to bear in mind going forward

**Pull request merged**